A minimal `engine_tools_lib` to use for local-repo Dart tooling

matanlurey opened the pull request:
Partial work towards re-landing #44936.

Both the `clang_tidy` and `githooks` passage could benefit from being able to automatically find the latest `compile_commands.json` output, which means that some common code should exist in the `tools/` directory.

This is a very minimal (but tested) library for doing exactly that.

zanderso commented on tools/engine_tools_lib/lib/engine_tools_lib.dart:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    // Check for the existence of a `out` directory within `src`.
+    final io.Directory outDir = io.Directory(p.join(srcPath, 'out'));
```
Could this be optional? `outputs()` below could return an empty list, and `latestOutput()` could return `null`.

matanlurey replied:
Yeah, very fair. Updated.

zanderso reviewed (commented):
I like this idea, but wondering if we shouldn't put this under a new subdirectory like `tools/pkg` (`pkg` to mirror subdirectories containing several Dart packages). Also wondering about the name. What do you think of `engine_repo_tools`?

matanlurey commented on tools/engine_tools_lib/lib/engine_tools_lib.dart:
```diff
@@ -0,0 +1,184 @@
+
+    do {
+      if (p.basename(maybeSrcDir.path) == 'src') {
+        return Engine.fromSrcPath(maybeSrcDir.path);
```
Great idea, updated.

matanlurey commented:
> I like this idea, but wondering if we shouldn't put this under a new subdirectory like `tools/pkg` (`pkg` to mirror subdirectories containing several Dart packages). Also wondering about the name. What do you think of `engine_repo_tools`?

Works for me! Moved and renamed.

zanderso commented on tools/pkg/engine_repo_tools/lib/engine_repo_tools.dart:
````diff
@@ -0,0 +1,228 @@
+  /// ```
+  ///
+  /// Throws a [InvalidEngineException] if the path is not a valid engine root.
+  factory Engine.fromSrcPath(String srcPath) {
````
Our conversation from lunch reminded me that passing a `FileSystem` from `package:file` here and to `findWithin` will let this library and its clients test with a `MemoryFileSystem` rather than the real thing.

matanlurey replied:
Yeah that's fair.

Here is a question though, is that particularly useful? For example, where this would be used (`clang_tidy`, `githooks`) there isn't a real way to make them work on a virtual file system - which means we'd be adding the abstraction to `Engine` for ... engine to test itself?

Dunno, seems sus - I'm not against it but I'd prefer to add it when it provides value?

zanderso replied:
Sure, we can add it later if we need it.
